File: ios/PolarBleSdkModule.swift
//
//  PolarBleSdkModule.swift
//  ReactNativePolarBleSDK


import Foundation
import PolarBleSdk
import RxSwift
import CoreBluetooth

@objc(PolarBleSdkModule)
class PolarBleSdkModule: NSObject {
  
  private var api = PolarBleApiDefaultImpl.polarImplementation(DispatchQueue.main, features: Features.allFeatures.rawValue)
  private var isSearchOn: Bool = false
  private var searchDisposable: Disposable?
  private let disposeBag = DisposeBag()
  
  override init() {
    super.init()
    api.polarFilter(true)
    api.observer = self
    api.deviceFeaturesObserver = self
    api.powerStateObserver = self
    api.deviceInfoObserver = self
    api.sdkModeFeatureObserver = self
    api.deviceHrObserver = self
    //api.logger = self
  }
  
  @objc(searchForDevice)
  func searchForDevice() -> Void {
    NSLog("native searchForDevice called")
    if !isSearchOn {
      isSearchOn = true
      searchDisposable = api.searchForDevice()
        .observe(on: MainScheduler.instance)
        .subscribe{ e in
          switch e {
          case .completed:
            NSLog("search complete")
            self.isSearchOn = false
          case .error(let err):
            NSLog("search error: \(err)")
            self.isSearchOn = false
          case .next(let item):
            NSLog("polar device found: \(item.name) connectable: \(item.connectable) address: \(item.address.uuidString)")
          }
        }
    } else {
      isSearchOn = false
      searchDisposable?.dispose()
    }
  }
  
  @objc
  func connectToDevice(_ deviceId: String) -> Void {
    do {
      try api.connectToDevice(deviceId)
    } catch let err {
      NSLog("Failed to connect to \(deviceId). Reason \(err)")
    }
  }
  
  @objc
  func startEcgStream(_ deviceId: String) -> Void {
    getStreamSettings(deviceId: deviceId, feature:  DeviceStreamingFeature.ecg)
      .asObservable()
      .flatMap({
        (settings) -> Observable<PolarEcgData> in
        return self.api.startEcgStreaming(deviceId, settings: settings)
      }).observe(on: MainScheduler.instance)
      .subscribe{ e in
        switch e {
        case .next(let data):
          for µv in data.samples {
            NSLog("ECG    µV: \(µv)")
          }
        case .error(let err):
          NSLog("ECG stream failed: \(err)")
        case .completed:
          NSLog("ECG stream completed")
        }
      }.disposed(by: disposeBag)
  }
  
  fileprivate func getStreamSettings(deviceId: String, feature: PolarBleSdk.DeviceStreamingFeature) -> Single<PolarSensorSetting> {
    NSLog("Stream settings fetch for \(feature)")
    return api.requestStreamSettings(deviceId, feature: feature)
      .map { settings -> PolarSensorSetting in settings.maxSettings() }
  }
  
  @objc
  func constantsToExport() -> [String: Any]! {
    return ["someKey": "someValue"]
  }
  
}

// MARK: - PolarBleApiPowerStateObserver
extension PolarBleSdkModule : PolarBleApiPowerStateObserver {
  func blePowerOn() {
    NSLog("BLE ON")
  }
  
  func blePowerOff() {
    NSLog("BLE OFF")
  }
}

// MARK: - PolarBleApiObserver
extension PolarBleSdkModule : PolarBleApiObserver {
  func deviceConnecting(_ polarDeviceInfo: PolarDeviceInfo) {
    NSLog("DEVICE CONNECTING: \(polarDeviceInfo)")
    
  }
  
  func deviceConnected(_ polarDeviceInfo: PolarDeviceInfo) {
    NSLog("DEVICE CONNECTED: \(polarDeviceInfo)")
    
  }
  
  func deviceDisconnected(_ polarDeviceInfo: PolarDeviceInfo) {
    NSLog("DISCONNECTED: \(polarDeviceInfo)")
  }
}

// MARK: - PolarBleApiDeviceInfoObserver
extension PolarBleSdkModule : PolarBleApiDeviceInfoObserver {
  func batteryLevelReceived(_ identifier: String, batteryLevel: UInt) {
    NSLog("battery level updated: \(batteryLevel)")
  }
  
  func disInformationReceived(_ identifier: String, uuid: CBUUID, value: String) {
    NSLog("dis info: \(uuid.uuidString) value: \(value)")
  }
}

// MARK: - PolarBleApiSdkModeFeatureObserver
extension PolarBleSdkModule : PolarBleApiDeviceFeaturesObserver {
  func hrFeatureReady(_ identifier: String) {
    NSLog("HR ready")
  }
  
  func ftpFeatureReady(_ identifier: String) {
    NSLog("FTP ready")
  }
  
  func streamingFeaturesReady(_ identifier: String, streamingFeatures: Set<DeviceStreamingFeature>) {
    for feature in streamingFeatures {
      NSLog("Feature \(feature) is ready.")
    }
  }
}

// MARK: - PolarBleApiSdkModeFeatureObserver
extension PolarBleSdkModule : PolarBleApiSdkModeFeatureObserver {
  func sdkModeFeatureAvailable(_ identifier: String) {
    NSLog("SDK mode feature available. Device \(identifier)")
  }
}

// MARK: - PolarBleApiDeviceHrObserver
extension PolarBleSdkModule : PolarBleApiDeviceHrObserver {
  func hrValueReceived(_ identifier: String, data: PolarHrData) {
    NSLog("(\(identifier)) HR value: \(data.hr) rrsMs: \(data.rrsMs) rrs: \(data.rrs) contact: \(data.contact) contact supported: \(data.contactSupported)")
  }
}

// MARK: - PolarBleApiLogger
extension PolarBleSdkModule : PolarBleApiLogger {
  func message(_ str: String) {
    NSLog("Polar SDK log:  \(str)")
  }
}

File: android/app/src/main/java/com/reactnativepolarblesdk/PolarBleSdkModule.kt
package com.reactnativepolarblesdk

import android.util.Log
import android.util.Pair
import com.facebook.react.bridge.ReactApplicationContext
import com.facebook.react.bridge.ReactContextBaseJavaModule
import com.facebook.react.bridge.ReactMethod
import com.polar.sdk.api.PolarBleApi
import com.polar.sdk.api.PolarBleApiCallback
import com.polar.sdk.api.PolarBleApiDefaultImpl
import com.polar.sdk.api.errors.PolarInvalidArgument
import com.polar.sdk.api.model.PolarDeviceInfo
import com.polar.sdk.api.model.PolarEcgData
import com.polar.sdk.api.model.PolarHrData
import com.polar.sdk.api.model.PolarSensorSetting
import io.reactivex.rxjava3.android.schedulers.AndroidSchedulers
import io.reactivex.rxjava3.core.Flowable
import io.reactivex.rxjava3.core.Single
import java.util.*

class PolarBleSdkModule(reactContext: ReactApplicationContext) : ReactContextBaseJavaModule(reactContext) {
    companion object {
        private const val TAG = "PolarBleSdkModule"
        private const val API_LOGGER_TAG = "API LOGGER"
    }

    private val api: PolarBleApi by lazy {
        // Notice PolarBleApi.ALL_FEATURES are enabled
        PolarBleApiDefaultImpl.defaultImplementation(reactContext, PolarBleApi.ALL_FEATURES)
    }

    init {
        api.setPolarFilter(false)
        api.setApiLogger { s: String -> Log.d(API_LOGGER_TAG, s) }
        api.setApiCallback(object : PolarBleApiCallback() {
            override fun blePowerStateChanged(powered: Boolean) {
                Log.d(TAG, "BLE power: $powered")
            }

            override fun deviceConnected(polarDeviceInfo: PolarDeviceInfo) {
                Log.d(TAG, "CONNECTED: " + polarDeviceInfo.deviceId)
            }

            override fun deviceConnecting(polarDeviceInfo: PolarDeviceInfo) {
                Log.d(TAG, "CONNECTING: " + polarDeviceInfo.deviceId)
            }

            override fun deviceDisconnected(polarDeviceInfo: PolarDeviceInfo) {
                Log.d(TAG, "DISCONNECTED: " + polarDeviceInfo.deviceId)
            }

            override fun streamingFeaturesReady(identifier: String, features: Set<PolarBleApi.DeviceStreamingFeature>) {
                for (feature in features) {
                    Log.d(TAG, "Streaming feature $feature is ready")
                }
            }

            override fun hrFeatureReady(identifier: String) {
                Log.d(TAG, "HR READY: $identifier")
                // hr notifications are about to start
            }

            override fun disInformationReceived(identifier: String, uuid: UUID, value: String) {
                Log.d(TAG, "uuid: $uuid value: $value")
            }

            override fun batteryLevelReceived(identifier: String, level: Int) {
                Log.d(TAG, "BATTERY LEVEL: $level")
            }

            override fun hrNotificationReceived(identifier: String, data: PolarHrData) {
                Log.d(TAG, "HR value: ${data.hr} rrsMs: ${data.rrsMs} rr: ${data.rrs} contact: ${data.contactStatus} , ${data.contactStatusSupported}")
            }

            override fun polarFtpFeatureReady(s: String) {
                Log.d(TAG, "FTP ready")
            }
        })

    }

    override fun getName(): String {
        return "PolarBleSdkModule"
    }

    @ReactMethod
    fun searchForDevice() {
        Log.d(TAG, "searchForDevice called")
        api.searchForDevice()
            .observeOn(AndroidSchedulers.mainThread())
            .subscribe(
                { polarDeviceInfo: PolarDeviceInfo ->
                    Log.d(TAG, "polar device found id: " + polarDeviceInfo.deviceId + " address: " + polarDeviceInfo.address + " rssi: " + polarDeviceInfo.rssi + " name: " + polarDeviceInfo.name + " isConnectable: " + polarDeviceInfo.isConnectable)
                },
                { error: Throwable ->
                    Log.e(TAG, "Device scan failed. Reason $error")
                },
                { Log.d(TAG, "complete") }
            )
    }

    @ReactMethod
    fun connectToDevice(deviceId: String) {
        Log.d(TAG, "connectToDevice called")
        try {
            api.connectToDevice(deviceId)
        } catch (polarInvalidArgument: PolarInvalidArgument) {
            Log.e(TAG, "Failed to connect. Reason $polarInvalidArgument ")
        }
    }

    @ReactMethod
    fun startEcgStream(deviceId: String) {
        Log.d(TAG, "startEcgStream called")
        requestStreamSettings(deviceId, PolarBleApi.DeviceStreamingFeature.ECG)
            .flatMap { settings: PolarSensorSetting ->
                api.startEcgStreaming(deviceId, settings)
            }
            .subscribe(
                { polarEcgData: PolarEcgData ->
                    for (microVolts in polarEcgData.samples) {
                        Log.d(TAG, "    yV: $microVolts")
                    }
                },
                { error: Throwable ->
                    Log.e(TAG, "ECG stream failed. Reason $error")
                },
                { Log.d(TAG, "ECG stream complete") }
            )
    }

    private fun requestStreamSettings(identifier: String, feature: PolarBleApi.DeviceStreamingFeature): Flowable<PolarSensorSetting> {
        val availableSettings = api.requestStreamSettings(identifier, feature)
            .observeOn(AndroidSchedulers.mainThread())
            .onErrorReturn { error: Throwable ->
                val errorString = "Settings are not available for feature $feature. REASON: $error"
                Log.w(TAG, errorString)
                PolarSensorSetting(emptyMap())
            }
        val allSettings = api.requestFullStreamSettings(identifier, feature)
            .onErrorReturn { error: Throwable ->
                Log.w(TAG, "Full stream settings are not available for feature $feature. REASON: $error")
                PolarSensorSetting(emptyMap())
            }
        return Single.zip(availableSettings, allSettings) { available: PolarSensorSetting, all: PolarSensorSetting ->
            if (available.settings.isEmpty()) {
                throw Throwable("Settings are not available")
            } else {
                Log.d(TAG, "Feature " + feature + " available settings " + available.settings)
                Log.d(TAG, "Feature " + feature + " all settings " + all.settings)
                return@zip Pair(available, all)
            }
        }
            .observeOn(AndroidSchedulers.mainThread())
            .toFlowable()
            .map { sensorSettings: Pair<PolarSensorSetting, PolarSensorSetting> ->
                sensorSettings.first.maxSettings()
            }
    }
}
